components/common-go/watch/file.go

```diff
@@ -0,0 +1,123 @@
+// Copyright (c) 2022 Gitpod GmbH. All rights reserved.
+// Licensed under the GNU Affero General Public License (AGPL).
+// See License-AGPL.txt in the project root for license information.
+
+package watch
+
+import (
+	"context"
+	"crypto/sha256"
+	"encoding/hex"
+	"io"
+	"os"
+	"path/filepath"
+
+	"github.com/fsnotify/fsnotify"
+	"golang.org/x/xerrors"
+
+	"github.com/gitpod-io/gitpod/common-go/log"
+)
+
+type fileWatcher struct {
+	onChange func()
+
+	watcher *fsnotify.Watcher
+
+	hash string
+}
+
+func File(ctx context.Context, path string, onChange func()) error {
+	watcher, err := fsnotify.NewWatcher()
+	if err != nil {
+		return xerrors.Errorf("unexpected error creating file watcher: %w", err)
+	}
+
+	fw := &fileWatcher{
+		watcher:  watcher,
+		onChange: onChange,
+	}
+
+	// initial hash of the file
+	hash, err := hashConfig(path)
+	if err != nil {
+		return xerrors.Errorf("cannot get hash of file %v: %w", path, err)
+	}
+
+	watchDir, _ := filepath.Split(path)
+	err = watcher.Add(watchDir)
```

Comment:
If I understand correctly, this watch the directory rather than the config path?

Reply:
Yes, this is a weird k8s thing. The keys in the configmap are not mapped to files but to a symlink. Because how fsnotify works, when we update the configmap, there's an atomic update that creates a new directory and swaps the mapped keys.
There was an attempt to fix this by adding the function fsnotify.AddRaw (watching the symlink) but that introduced some weird behaviors. 
For that reasons we watch the directory :(